torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -0,0 +1,390 @@
+    case ReduceOp::PRODUCT:
+      return ::gloo::ReductionFunction<T>::product;
+    case ReduceOp::MIN:
+      return ::gloo::ReductionFunction<T>::min;
+    case ReduceOp::MAX:
+      return ::gloo::ReductionFunction<T>::max;
+  }
+
+  throw std::runtime_error("Unhandled ReduceOp");
+}
+
+} // namespace
+
+ProcessGroupGloo::WorkGloo::WorkGloo() {
+  completed_ = false;
+}
+
+ProcessGroupGloo::WorkGloo::~WorkGloo() {
+}
+
+bool ProcessGroupGloo::WorkGloo::isCompleted() {
+  std::unique_lock<std::mutex> lock(m_);
+  return completed_;
+}
+
+bool ProcessGroupGloo::WorkGloo::wait() {
+  std::unique_lock<std::mutex> lock(m_);
+  while (!completed_) {
+    cv_.wait(lock);
+  }
+  return !ex_;
+}
+
+const std::exception& ProcessGroupGloo::WorkGloo::exception() const {
+  return *ex_;
+}
+
+void ProcessGroupGloo::WorkGloo::finish() {
+  {
+    std::unique_lock<std::mutex> lock(m_);
+    completed_ = true;
+  }
+  cv_.notify_all();
+}
+
+void ProcessGroupGloo::WorkGloo::finishWithException(
+    const ::gloo::Exception& ex) {
+  {
+    std::unique_lock<std::mutex> lock(m_);
+    completed_ = true;
+    ex_ = std::unique_ptr<::gloo::Exception>(new ::gloo::Exception(ex));
+  }
+  cv_.notify_all();
+}
+
+ProcessGroupGloo::ProcessGroupGloo(
+    const std::shared_ptr<Store>& store,
+    const std::vector<std::shared_ptr<::gloo::transport::Device>>& devices,
+    int rank,
+    int size)
+    : ProcessGroup(rank, size),
+      store_(new GlooStore(store)),
+      devices_(devices) {
+  stop_ = false;
+}
+
+ProcessGroupGloo::~ProcessGroupGloo() {
+  // Require this process group to be explicitly shut down prior to being
+  // destructed. This is the easiest way to guarantee clean shutdown
+  // and avoid blocking/throwing in a destructor.
+}
+
+void ProcessGroupGloo::initialize() {
+  contexts_.clear();
+  for (auto& device : devices_) {
+    auto context = std::shared_ptr<::gloo::rendezvous::Context>(
+        new ::gloo::rendezvous::Context(rank_, size_));
+    context->setTimeout(std::chrono::milliseconds(100));
+    context->connectFullMesh(*store_, device);
+    contexts_.push_back(std::move(context));
+  }
+
+  // TODO(@pietern): Make number of threads configurable
+  threads_.resize(2);
+  for (int i = 0; i < threads_.size(); i++) {
+    threads_[i] = std::thread(&ProcessGroupGloo::runLoop, this);
+  }
+}
+
+void ProcessGroupGloo::destroy() {
+  std::unique_lock<std::mutex> lock(m_);
+  while (!queue_.empty()) {
+    queueConsumeCV_.wait(lock);
+  }
+
+  // Queue is empty, signal stop
+  stop_ = true;
+
+  // Release lock to allow threads to terminate
+  queueProduceCV_.notify_all();
+  lock.unlock();
+
+  // Wait for worker threads to terminate
+  for (auto& thread : threads_) {
+    thread.join();
+  }
+}
+
+void ProcessGroupGloo::runLoop(void) {
+  std::unique_lock<std::mutex> lock(m_);
+
+  while (!stop_) {
+    if (queue_.empty()) {
+      queueProduceCV_.wait(lock);
+      continue;
+    }
+
+    auto tuple = std::move(queue_.front());
```

Comment:
Can also use std::tie to unpack this tuple directly instead of getting them one by one

Reply:
That means you have to declare them first right?

Reply:
Yes, in a different namespace tcputil, just a thought, not really a big deal.